File: 081. Search in Rotated Sorted Array II/Java/Solution.java
public class Solution {
    public boolean search(int[] nums, int target) {
        int low = 0;
        int high = nums.length - 1;
        while (low <= high) {
            while (low < high && nums[low] == nums[high]) {
                low++;
            }
            int mid = (low + high) / 2;
            if (nums[mid] == target) {
                return true;
            }
            if (nums[mid] >= nums[low]) {
                if (nums[mid] < target) {
                    low = mid + 1;
                } else {
                    if (nums[low] > target) {
                        low = mid + 1;
                    } else {
                        high = mid - 1;
                    }
                }
            } else {
                if (nums[mid] > target) {
                    high = mid - 1;
                } else {
                    if (nums[high] < target) {
                        high = mid - 1;
                    } else {
                        low = mid + 1;
                    }
                }
            }
        }
        return false;
    }
}

File: 081. Search in Rotated Sorted Array II/Python/Solution.py
class Solution(object):
    def search(self, nums, target):
        """
        :type nums: List[int]
        :type target: int
        :rtype: bool
        """
        if not nums:
            return False
        low = 0
        high = len(nums) - 1
        while low < high:
            mid = (low + high) // 2
            if nums[mid] == target:
                return True
            if nums[mid] > nums[high]:
                if nums[low] <= target < nums[mid]:
                    high = mid
                else:
                    low = mid + 1
            elif nums[mid] < nums[high]:
                if nums[mid] < target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid
            else:
                high -= 1
        return nums[low] == target
